stdlib/random.ml

```diff
@@ -13,76 +14,69 @@
 (*                                                                        *)
 (**************************************************************************)
 
-(* Pseudo-random number generator
-   This is a lagged-Fibonacci F(55, 24, +) with a modified addition
-   function to enhance the mixing of bits.
-   If we use normal addition, the low-order bit fails tests 1 and 7
-   of the Diehard test suite, and bits 1 and 2 also fail test 7.
-   If we use multiplication as suggested by Marsaglia, it doesn't fare
-   much better.
-   By mixing the bits of one of the numbers before addition (XOR the
-   5 high-order bits into the low-order bits), we get a generator that
-   passes all the Diehard tests.
-*)
+(* Pseudo-random number generator *)
 
 external random_seed: unit -> int array = "caml_sys_random_seed"
 
 module State = struct
 
-  type t = { st : int array; mutable idx : int }
+  open Bigarray
 
-  let new_state () = { st = Array.make 55 0; idx = 0 }
-  let assign st1 st2 =
-    Array.blit st2.st 0 st1.st 0 55;
-    st1.idx <- st2.idx
+  type t = (int64, int64_elt, c_layout) Array1.t
 
+  external next: t -> (int64[@unboxed])
+      = "caml_lxm_next" "caml_lxm_next_unboxed" [@@noalloc]
 
-  let full_init s seed =
-    let combine accu x = Digest.string (accu ^ Int.to_string x) in
-    let extract d =
-      Char.code d.[0] + (Char.code d.[1] lsl 8) + (Char.code d.[2] lsl 16)
-      + (Char.code d.[3] lsl 24)
-    in
-    let seed = if Array.length seed = 0 then [| 0 |] else seed in
-    let l = Array.length seed in
-    for i = 0 to 54 do
-      s.st.(i) <- i;
-    done;
-    let accu = ref "x" in
-    for i = 0 to 54 + Int.max 55 l do
-      let j = i mod 55 in
-      let k = i mod l in
-      accu := combine !accu seed.(k);
-      s.st.(j) <- (s.st.(j) lxor extract !accu) land 0x3FFFFFFF;  (* PR#5575 *)
-    done;
-    s.idx <- 0
+  let create () : t =
+    Array1.create Int64 C_layout 4
 
+  let set s i1 i2 i3 i4 =
+    Array1.unsafe_set s 0 (Int64.logor i1 1L); (* must be odd *)
+    Array1.unsafe_set s 1 i2;
+    Array1.unsafe_set s 2 (if i3 <> 0L then i3 else 1L); (* must not be 0 *)
+    Array1.unsafe_set s 3 (if i4 <> 0L then i4 else 2L) (* must not be 0 *)
```

Comment:
The reason why those `unsafe_` are correct is thanks to type abstraction: the module exports an abstract state type `t`, and the three exported functions that can build a `t` (`make`, `make_self_init`, `copy`) guarantee that the size is exactly 4.